File: vertx-mini-api/src/main/java/tk/okou/vertx/sdk/tencent/qq/QQMiniGameApi.java
package tk.okou.vertx.sdk.tencent.qq;

import io.vertx.codegen.annotations.Fluent;
import io.vertx.codegen.annotations.Nullable;
import io.vertx.codegen.annotations.VertxGen;
import io.vertx.core.AsyncResult;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;
import io.vertx.core.json.JsonObject;
import tk.okou.vertx.sdk.BaseMiniGameApi;
import tk.okou.vertx.sdk.tencent.qq.impl.QQMiniGameApiImpl;
import tk.okou.vertx.sdk.model.KVData;
import tk.okou.sdk.util.SignatureMethod;

import java.util.List;

@VertxGen
public interface QQMiniGameApi extends BaseMiniGameApi, QQMiniApi {
    static QQMiniGameApi create(Vertx vertx, QQMiniApiOptions options) {
        return new QQMiniGameApiImpl(vertx, options);
    }

    static QQMiniGameApi create(Vertx vertx) {
        return new QQMiniGameApiImpl(vertx, new QQMiniApiOptions());
    }
    @Fluent
    QQMiniGameApi code2session(String appId, String secret, String jsCode, Handler<AsyncResult<JsonObject>> handler);

    @Fluent
    QQMiniGameApi code2session(String appId, String secret, String jsCode, String grantType, Handler<AsyncResult<JsonObject>> handler);

    @Fluent
    QQMiniGameApi getAccessToken(String appId, String secret, Handler<AsyncResult<JsonObject>> handler);

    @Fluent
    QQMiniGameApi getAccessToken(String grantType, String appId, String secret, Handler<AsyncResult<JsonObject>> handler);

    @Fluent
    QQMiniGameApi setUserStorage(String accessToken, String openId, String sessionKey, List<KVData> kvList, Handler<AsyncResult<JsonObject>> handler);

    @Fluent
    QQMiniGameApi setUserStorage(String accessToken, String openId, String sessionKey, SignatureMethod signatureMethod, List<KVData> kvList, Handler<AsyncResult<JsonObject>> handler);

    @Fluent
    QQMiniGameApi removeUserStorage(String accessToken, String openId, String sessionKey, List<String> keys, Handler<AsyncResult<JsonObject>> handler);

    @Fluent
    QQMiniGameApi removeUserStorage(String accessToken, String openId, String sessionKey, SignatureMethod signatureMethod, List<String> key, Handler<AsyncResult<JsonObject>> handler);

    @Fluent
    QQMiniGameApi sendSubscriptionMessage(String accessToken, String toUserOpenId, String templateId, @Nullable String page, @Nullable JsonObject data, @Nullable String emphasisKeyWord, @Nullable String oacAppId, int useRobot, Handler<AsyncResult<JsonObject>> handler);

}

File: vertx-mini-api-kotlin-coroutines/src/main/kotlin/tk/okou/vertx/sdk/kotlin/tencent/qq/QQMiniGameApi.kt
/*
 * Copyright 2019 Red Hat, Inc.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * and Apache License v2.0 which accompanies this distribution.
 *
 * The Eclipse Public License is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * The Apache License v2.0 is available at
 * http://www.opensource.org/licenses/apache2.0.php
 *
 * You may elect to redistribute this code under either of these licenses.
 */
package tk.okou.vertx.sdk.kotlin.tencent.qq

import io.vertx.core.json.JsonObject
import io.vertx.kotlin.coroutines.awaitResult
import tk.okou.sdk.util.SignatureMethod
import tk.okou.vertx.sdk.model.KVData
import tk.okou.vertx.sdk.tencent.qq.QQMiniGameApi

@Deprecated(message = "Instead use code2session returning a future and chain with await()", replaceWith = ReplaceWith("code2session(appId, secret, jsCode).await()"))
suspend fun QQMiniGameApi.code2sessionAwait(appId: String, secret: String, jsCode: String): JsonObject {
  return awaitResult {
    this.code2session(appId, secret, jsCode, it)
  }
}

@Deprecated(message = "Instead use code2session returning a future and chain with await()", replaceWith = ReplaceWith("code2session(appId, secret, jsCode, grantType).await()"))
suspend fun QQMiniGameApi.code2sessionAwait(appId: String, secret: String, jsCode: String, grantType: String): JsonObject {
  return awaitResult {
    this.code2session(appId, secret, jsCode, grantType, it)
  }
}

@Deprecated(message = "Instead use getAccessToken returning a future and chain with await()", replaceWith = ReplaceWith("getAccessToken(appId, secret).await()"))
suspend fun QQMiniGameApi.getAccessTokenAwait(appId: String, secret: String): JsonObject {
  return awaitResult {
    this.getAccessToken(appId, secret, it)
  }
}

@Deprecated(message = "Instead use getAccessToken returning a future and chain with await()", replaceWith = ReplaceWith("getAccessToken(grantType, appId, secret).await()"))
suspend fun QQMiniGameApi.getAccessTokenAwait(grantType: String, appId: String, secret: String): JsonObject {
  return awaitResult {
    this.getAccessToken(grantType, appId, secret, it)
  }
}

@Deprecated(message = "Instead use setUserStorage returning a future and chain with await()", replaceWith = ReplaceWith("setUserStorage(accessToken, openId, sessionKey, kvList).await()"))
suspend fun QQMiniGameApi.setUserStorageAwait(accessToken: String, openId: String, sessionKey: String, kvList: List<KVData>): JsonObject {
  return awaitResult {
    this.setUserStorage(accessToken, openId, sessionKey, kvList, it)
  }
}

@Deprecated(message = "Instead use setUserStorage returning a future and chain with await()", replaceWith = ReplaceWith("setUserStorage(accessToken, openId, sessionKey, signatureMethod, kvList).await()"))
suspend fun QQMiniGameApi.setUserStorageAwait(accessToken: String, openId: String, sessionKey: String, signatureMethod: SignatureMethod, kvList: List<KVData>): JsonObject {
  return awaitResult {
    this.setUserStorage(accessToken, openId, sessionKey, signatureMethod, kvList, it)
  }
}

@Deprecated(message = "Instead use removeUserStorage returning a future and chain with await()", replaceWith = ReplaceWith("removeUserStorage(accessToken, openId, sessionKey, keys).await()"))
suspend fun QQMiniGameApi.removeUserStorageAwait(accessToken: String, openId: String, sessionKey: String, keys: List<String>): JsonObject {
  return awaitResult {
    this.removeUserStorage(accessToken, openId, sessionKey, keys, it)
  }
}

@Deprecated(message = "Instead use removeUserStorage returning a future and chain with await()", replaceWith = ReplaceWith("removeUserStorage(accessToken, openId, sessionKey, signatureMethod, key).await()"))
suspend fun QQMiniGameApi.removeUserStorageAwait(accessToken: String, openId: String, sessionKey: String, signatureMethod: SignatureMethod, key: List<String>): JsonObject {
  return awaitResult {
    this.removeUserStorage(accessToken, openId, sessionKey, signatureMethod, key, it)
  }
}

